File: src/promptWithExtendedThinking.ts
#!/usr/bin/env ts-node

/**
 * Extended Thinking Example for Claude 3.7 Sonnet
 *
 * This script demonstrates basic extended thinking capabilities of Claude 3.7 Sonnet.
 * It allows you to set a thinking budget to improve Claude's reasoning on complex tasks.
 *
 * Usage:
 *    npm run start:prompt-with-extended-thinking -- --prompt "Explain quantum computing to me" --max-tokens 2048 --thinking-budget-tokens 1024
 */

import * as dotenv from 'dotenv';
import { program } from 'commander';
import Anthropic from '@anthropic-ai/sdk';
import type { TextBlock } from '@anthropic-ai/sdk/resources/messages';
import chalk from 'chalk';
import Table from 'cli-table3';
import ora from 'ora';
import { ThinkingBlock, ExtendedMessageCreateParams, ExtendedMessage } from './types';

// Extend the TableConstructorOptions interface to include the title property
declare module 'cli-table3' {
  interface TableConstructorOptions {
    title?: string;
  }
}

// Load environment variables from .env file
dotenv.config();

async function main() {
  // Parse command line arguments
  program
    .description('Claude 3.7 Sonnet extended thinking example')
    .requiredOption('--prompt <prompt>', 'The prompt to send to Claude')
    .option('--max-tokens <tokens>', 'Maximum number of tokens in the response', '2000')
    .option('--thinking-budget-tokens <tokens>', 'Budget for thinking tokens (minimum 1024)', '8000');

  program.parse(process.argv);
  const options = program.opts<{
    prompt: string;
    maxTokens: string;
    thinkingBudgetTokens: string;
  }>();

  // Get API key from environment variable
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    console.error(chalk.bold.red('Error:'), 'ANTHROPIC_API_KEY environment variable not set');
    process.exit(1);
  }

  // Convert tokens to numbers and validate
  let thinkingBudget = parseInt(options.thinkingBudgetTokens, 10);
  let maxTokens = parseInt(options.maxTokens, 10);

  // Validate thinking budget (minimum 1024 tokens)
  if (thinkingBudget < 1024) {
    console.warn(chalk.bold.yellow('Warning:'), 'Minimum thinking budget is 1024 tokens. Setting to 1024.');
    thinkingBudget = 1024;
  }

  // Ensure max_tokens is greater than thinking_budget_tokens
  if (maxTokens <= thinkingBudget) {
    const newMaxTokens = thinkingBudget + 1000;
    console.warn(chalk.bold.yellow('Warning:'), `max-tokens must be greater than thinking-budget-tokens. Setting max-tokens to ${newMaxTokens}.`);
    maxTokens = newMaxTokens;
  }

  // Initialize the Anthropic client
  const client = new Anthropic({
    apiKey,
  });

  try {
    // Display request information
    console.log(
      chalk.green.bold('\n===== Request to Claude 3.7 Sonnet =====\n') +
      `${chalk.bold('Prompt:')} ${options.prompt}\n` +
      `${chalk.bold('Max Tokens:')} ${maxTokens}\n` +
      `${chalk.bold('Thinking Budget:')} ${thinkingBudget} tokens\n`
    );

    // Create a message with Claude 3.7 Sonnet using extended thinking
    const spinner = ora(`Sending request to Claude 3.7 Sonnet with ${thinkingBudget} thinking tokens...`).start();
    
    const response = await client.messages.create({
      model: 'claude-3-7-sonnet-20250219',
      max_tokens: maxTokens,
      thinking: {
        type: 'enabled',
        budget_tokens: thinkingBudget,
      },
      messages: [{ role: 'user', content: options.prompt }],
    } as ExtendedMessageCreateParams) as unknown as ExtendedMessage;
    
    spinner.stop();

    // Log the API response
    console.log(
      chalk.green.bold('\n===== API Response =====\n') +
      JSON.stringify(response, null, 2) + '\n'
    );

    // Extract and display the thinking block if present
    const thinkingBlocks = response.content
      .filter((block): block is ThinkingBlock => block.type === 'thinking' as any);
    
    if (thinkingBlocks.length > 0) {
      console.log(
        chalk.cyan.bold('\n===== Claude\'s Thinking Process =====\n') +
        thinkingBlocks[0].thinking + '\n'
      );
    }

    // Display the text response
    const textBlocks = response.content
      .filter((block): block is TextBlock => block.type === 'text');
    
    if (textBlocks.length > 0) {
      console.log(
        chalk.blue.bold('\n===== Claude\'s Response =====\n') +
        textBlocks[0].text + '\n'
      );

      // Calculate approximate token usage
      // This is a rough approximation. For accurate token counting, use a tokenizer
      const estimateTokens = (text: string): number => {
        // A better approximation than just splitting by spaces
        return Math.ceil(text.length / 4);
      };
      
      const promptTokens = estimateTokens(options.prompt);
      const thinkingTokens = thinkingBlocks.length > 0 ? estimateTokens(thinkingBlocks[0].thinking) : 0;
      const responseTokens = estimateTokens(textBlocks[0].text);
      const totalTokens = promptTokens + thinkingTokens + responseTokens;

      // Calculate approximate costs (Claude 3.7 Sonnet pricing)
      const inputCost = promptTokens * (3.0 / 1000000);  // $3.00 per million tokens
      const outputCost = responseTokens * (15.0 / 1000000);  // $15.00 per million tokens
      const thinkingCost = thinkingTokens * (15.0 / 1000000);  // $15.00 per million tokens (thinking tokens are billed as output)
      const totalCost = inputCost + outputCost + thinkingCost;

      // Display token usage summary
      const tokenTable = new Table({
        head: [chalk.cyan('Type'), chalk.magenta('Count'), chalk.green('Cost ($)')],
        title: 'Token Usage Summary'
      });

      tokenTable.push(
        ['Input Tokens', promptTokens.toString(), `$${inputCost.toFixed(6)}`],
        ['Output Tokens', responseTokens.toString(), `$${outputCost.toFixed(6)}`],
        ['Thinking Tokens', thinkingTokens.toString(), `$${thinkingCost.toFixed(6)}`],
        ['Total', totalTokens.toString(), `$${totalCost.toFixed(6)}`]
      );

      console.log(tokenTable.toString());
    }

  } catch (error) {
    console.error(chalk.bold.red('Error:'), error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

// Run the main function
main();

File: claude-3-7-sonnet-starter-pack/prompt_with_extended_thinking.py
#!/usr/bin/env -S uv run --script

# /// script
# dependencies = [
#   "anthropic>=0.45.2",
#   "argparse>=1.4.0",
#   "rich>=13.7.0",
# ]
# ///

"""
Extended Thinking Example for Claude 3.7 Sonnet

This script demonstrates basic extended thinking capabilities of Claude 3.7 Sonnet.
It allows you to set a thinking budget to improve Claude's reasoning on complex tasks.

Usage:
    uv run prompt_with_extended_thinking.py --prompt "Explain quantum computing to me" --max_tokens 2048 --thinking_budget_tokens 1024
"""

import os
import sys
import json
import argparse
import anthropic
from rich.console import Console
from rich.panel import Panel
from rich.syntax import Syntax
from rich.table import Table
from rich.progress import Progress, SpinnerColumn, TextColumn


def main():
    # Initialize rich console
    console = Console()

    parser = argparse.ArgumentParser(
        description="Claude 3.7 Sonnet extended thinking example"
    )
    parser.add_argument(
        "--prompt", type=str, required=True, help="The prompt to send to Claude"
    )
    parser.add_argument(
        "--max_tokens",
        type=int,
        default=2000,
        help="Maximum number of tokens in the response",
    )
    parser.add_argument(
        "--thinking_budget_tokens",
        type=int,
        default=8000,
        help="Budget for thinking tokens (minimum 1024)",
    )
    args = parser.parse_args()

    # Validate thinking budget (minimum 1024 tokens)
    if args.thinking_budget_tokens < 1024:
        console.print(
            "[bold yellow]Warning:[/bold yellow] Minimum thinking budget is 1024 tokens. Setting to 1024."
        )
        args.thinking_budget_tokens = 1024

    # Ensure max_tokens is greater than thinking_budget_tokens
    if args.max_tokens <= args.thinking_budget_tokens:
        console.print(
            f"[bold yellow]Warning:[/bold yellow] max_tokens must be greater than thinking_budget_tokens. Setting max_tokens to {args.thinking_budget_tokens + 1000}."
        )
        args.max_tokens = args.thinking_budget_tokens + 1000

    # Get API key from environment variable
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        console.print(
            "[bold red]Error:[/bold red] ANTHROPIC_API_KEY environment variable not set"
        )
        sys.exit(1)

    # Initialize the Anthropic client
    client = anthropic.Anthropic(api_key=api_key)

    try:
        # Display request information
        console.print(
            Panel(
                f"[bold]Prompt:[/bold] {args.prompt}\n[bold]Max Tokens:[/bold] {args.max_tokens}\n[bold]Thinking Budget:[/bold] {args.thinking_budget_tokens} tokens",
                title="[bold green]Request to Claude 3.7 Sonnet[/bold green]",
                border_style="green",
            )
        )

        # Create a message with Claude 3.7 Sonnet using extended thinking
        with console.status(
            f"[bold green]Sending request to Claude 3.7 Sonnet with {args.thinking_budget_tokens} thinking tokens...",
            spinner="dots",
        ):
            response = client.messages.create(
                model="claude-3-7-sonnet-20250219",
                max_tokens=args.max_tokens,
                thinking={
                    "type": "enabled",
                    "budget_tokens": args.thinking_budget_tokens,
                },
                messages=[{"role": "user", "content": args.prompt}],
            )

        # Log the API response
        console.print(
            Panel(
                Syntax(
                    json.dumps(response.model_dump(), indent=2),
                    "json",
                    theme="monokai",
                    word_wrap=True,
                ),
                title="[bold green]API Response[/bold green]",
                border_style="green",
            )
        )

        # Extract and display the thinking block if present
        thinking_blocks = [
            block for block in response.content if block.type == "thinking"
        ]
        if thinking_blocks:
            # Use syntax highlighting for the thinking process
            console.print(
                Panel(
                    Syntax(
                        thinking_blocks[0].thinking,
                        "markdown",
                        theme="monokai",
                        word_wrap=True,
                    ),
                    title="[bold cyan]Claude's Thinking Process[/bold cyan]",
                    border_style="cyan",
                )
            )

        # Display the text response
        text_blocks = [block for block in response.content if block.type == "text"]
        if text_blocks:
            console.print(
                Panel(
                    text_blocks[0].text,
                    title="[bold blue]Claude's Response[/bold blue]",
                    border_style="blue",
                )
            )

            # Calculate approximate token usage
            prompt_tokens = len(args.prompt.split())
            thinking_tokens = sum(
                len(block.thinking.split())
                for block in response.content
                if block.type == "thinking"
            )
            response_tokens = sum(
                len(block.text.split())
                for block in response.content
                if block.type == "text"
            )
            total_tokens = prompt_tokens + thinking_tokens + response_tokens

            # Calculate approximate costs (Claude 3.7 Sonnet pricing)
            input_cost = prompt_tokens * (3.0 / 1000000)  # $3.00 per million tokens
            output_cost = response_tokens * (
                15.0 / 1000000
            )  # $15.00 per million tokens
            thinking_cost = thinking_tokens * (
                15.0 / 1000000
            )  # $15.00 per million tokens (thinking tokens are billed as output)
            total_cost = input_cost + output_cost + thinking_cost

            # Display token usage summary
            token_table = Table(title="Token Usage Summary")
            token_table.add_column("Type", style="cyan")
            token_table.add_column("Count", style="magenta")
            token_table.add_column("Cost ($)", style="green")
            token_table.add_row(
                "Input Tokens", str(prompt_tokens), f"${input_cost:.6f}"
            )
            token_table.add_row(
                "Output Tokens", str(response_tokens), f"${output_cost:.6f}"
            )
            token_table.add_row(
                "Thinking Tokens", str(thinking_tokens), f"${thinking_cost:.6f}"
            )
            token_table.add_row("Total", str(total_tokens), f"${total_cost:.6f}")
            console.print(token_table)

    except Exception as e:
        console.print(f"[bold red]Error:[/bold red] {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    main()
